### PR TITLE
Typecheck cleanup

### PR DESCRIPTION
**Pull request by comonoidial**
Fixed the pretty printing of types with repect to parentheses and type variables.
Removed uselss GenSet and moved the typechecker to type package.

### CODE REVIEW

File: Code/src/main/java/nl/utwente/group10/haskell/type/Type.java [hunk 1]
```diff
@@ -30,7 +30,15 @@
     /**
      * @return The Haskell (type) representation of this type.
      */
-    public abstract String toHaskellType();
+    public String toHaskellType() {
+    	return this.toHaskellType(0);
+    }
+
+    /**
+     * @param The fixity of the context the type is shown in.
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/haskell/type/Type.java]: Could perhaps use a short definition, as well as a reference of what the values mean (i.e. that they should always be between 0 and 9).

File: Code/src/main/java/nl/utwente/group10/haskell/type/ConstT.java [hunk 2]
```diff
@@ -54,15 +54,18 @@ public final Type prune() {
...
         out.append(this.constructor);
 
         for (Type arg : this.args) {
             out.append(" ");
-            out.append(arg.toHaskellType());
+            out.append(arg.toHaskellType(10));
         }
 
+        if (fixity > 9 && this.args.length > 0)
+        	return "(" + out.toString() + ")";
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/haskell/type/ConstT.java]: Code style elsewhere is braces `{}` even for one-liners.

File: Code/src/main/java/nl/utwente/group10/haskell/type/TypeChecker.java [hunk 3]
```diff
@@ -7,34 +7,30 @@
...
 import nl.utwente.group10.haskell.exceptions.HaskellTypeError;
 import nl.utwente.group10.haskell.expr.Expr;
-import nl.utwente.group10.haskell.type.ConstT;
-import nl.utwente.group10.haskell.type.Type;
-import nl.utwente.group10.haskell.type.TypeClass;
-import nl.utwente.group10.haskell.type.VarT;
 
 /**
- * Implementation of the Hindley-Milner type system for Haskell types and expressions.
+ * Implementation of a typechecker for a simplified variant of Haskell.
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/haskell/type/TypeChecker.java]: Brutal, but honest. :+1:

File: Code/src/main/java/nl/utwente/group10/haskell/type/TypeChecker.java [hunk 4]
```diff
@@ -7,34 +7,30 @@
...
      * Logger for this class.
      */
-    private static final Logger logger = Logger.getLogger(HindleyMilner.class.getName());
+    private static final Logger logger = Logger.getLogger(TypeChecker.class.getName());
 
     /**
      * Offset for the creation of variable types.
      */
-    static int tvOffset = 0;
+    static int tvOffset = -1;
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/haskell/type/TypeChecker.java]: Why `-1`?

File: Code/src/main/java/nl/utwente/group10/haskell/type/VarT.java [hunk 5]
```diff
@@ -132,12 +148,12 @@ public VarT getFresh() {
             instance = this.instance.get().getFresh();
         }
 
-        return new VarT(this.name, this.constraints, instance);
+        return new VarT(this.prefix, this.uid, this.constraints, instance);
```
Comment (wandernauta) [inline Code/src/main/java/nl/utwente/group10/haskell/type/VarT.java]: If you return a new VarT with the same uid, is that uid still a unique identifier?

**Comment by wandernauta** [comment]
Other than these little things, this lgtm :+1: The pretty printing really does look much nicer and less alien this way. Thanks! :sparkles:

**Comment by comonoidial** [comment]
The concept of unique identifier is indeed broken, and is unfornately not easy to fix.
The problem is in displaying type variables to the user, and this depends how define the scope in which they need to be unique in the UI.
It deserves an issue report on its own to define scope in the UI and how that could be implemented.
This pull request will make type variables a little less broken, and making typevars more unique will break unit tests.

**Comment by wandernauta** [comment]
Understood, I'll just go ahead and merge this PR as is. Thanks again!

**Comment by wandernauta** [comment]
Merged in 1d5da88601f08375152a48261268196506eefe16.

**Pull request closed without merging**